crypto/pkcs7/pkcs7.c

```diff
@@ -1086,3 +1061,347 @@ OPENSSL_END_ALLOW_DEPRECATED
+  if (ctx == NULL || !EVP_PKEY_decrypt_init(ctx)) {
+    goto err;
+  }
+  size_t len;
+  if (!EVP_PKEY_decrypt(ctx, NULL, &len, ri->enc_key->data,
+                        ri->enc_key->length) ||
+      (ek = OPENSSL_malloc(len)) == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_EVP_LIB);
+    return 0;
+  }
+
+  int ok =
+      EVP_PKEY_decrypt(ctx, ek, &len, ri->enc_key->data, ri->enc_key->length);
+  // We only return 0 on critical failure, not decryption failure. However, we
+  // still need to set |ek| to NULL to signal decryption failure to callers so
+  // they can use random bytes as content encryption key for MMA defense.
+  if (!ok) {
+    OPENSSL_free(ek);
+    ek = NULL;
+  }
+
+  ret = 1;
+  *ek_out = ek;
+
+err:
+  EVP_PKEY_CTX_free(ctx);
+  return ret;
+}
+
+// pkcs7_cmp_ri is a comparison function, so it returns 0 if |ri| and |pcert|
+// match and 1 if they do not.
+static int pkcs7_cmp_ri(PKCS7_RECIP_INFO *ri, X509 *pcert) {
+  if (ri == NULL || ri->issuer_and_serial == NULL || pcert == NULL) {
+    return 1;
+  }
+  int ret =
+      X509_NAME_cmp(ri->issuer_and_serial->issuer, X509_get_issuer_name(pcert));
+  if (ret) {
+    return ret;
+  }
+  return ASN1_INTEGER_cmp(X509_get0_serialNumber(pcert),
+                          ri->issuer_and_serial->serial);
+}
+
+static BIO *pkcs7_data_decode(PKCS7 *p7, EVP_PKEY *pkey, X509 *pcert) {
+  GUARD_PTR(p7);
+  GUARD_PTR(pkey);
+  BIO *out = NULL, *cipher_bio = NULL, *data_bio = NULL;
+  ASN1_OCTET_STRING *data_body = NULL;
+  const EVP_CIPHER *cipher = NULL;
+  X509_ALGOR *enc_alg = NULL;
+  STACK_OF(PKCS7_RECIP_INFO) *rsk = NULL;
+  PKCS7_RECIP_INFO *ri = NULL;
+  uint8_t *cek = NULL, *dummy_key = NULL;  // cek means "content encryption key"
+
+  if (p7->d.ptr == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+    return NULL;
+  }
+
+  switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_enveloped:
+      rsk = p7->d.enveloped->recipientinfo;
+      enc_alg = p7->d.enveloped->enc_data->algorithm;
+      // |data_body| is NULL if the optional EncryptedContent is missing.
+      data_body = p7->d.enveloped->enc_data->enc_data;
+      cipher = EVP_get_cipherbynid(OBJ_obj2nid(enc_alg->algorithm));
+      if (cipher == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNSUPPORTED_CIPHER_TYPE);
+        goto err;
+      }
+      break;
+    default:
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNSUPPORTED_CONTENT_TYPE);
+      goto err;
+  }
+
+  if ((cipher_bio = BIO_new(BIO_f_cipher())) == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, ERR_R_BIO_LIB);
+    goto err;
+  }
+
+  // RFC 3218 provides an overview of, and mitigations for, the "Million Message
+  // Attack" (MMA) on RSA encryption with PKCS-1 padding. Section 2.3 describes
+  // implementor countermeasures. We implement the following countermeasures, as
+  // does OpenSSL.
+  //
+  // 1. Do not branch on |cek| decryption failure when checking recip infos
+  // 2. Clear error state after |cek| decrypt is attempted
```

Comment:
why would we even fill the error state when decrypting |cek|?

Reply:
that's below our level of control. the EVP API will populate an error on RSA decrypt failure (i.e. invalid PKCS #1.5 padding exploited in the MMA attack).